smart-contracts/contracts/PublicLock.sol

```diff
@@ -599,6 +649,68 @@ contract PublicLock is ILockCore, ERC165, IERC721, IERC721Receiver, Ownable {
     );
   }
 
+  /**
+   * @dev Destroys the user's key and sends a refund based on the amount of time remaining.
+   * This assumes the caller is the owner or is approved to modify the owner's key.
+   * @param _owner The owner of the key to be destroyed.
+   */
+  function _cancelAndRefund(
+    address _owner
+  )
+    internal
+  {
+    Key storage key = keyByOwner[_owner];
+
+    // 1) Calc refund
+    uint refund = _getCancelAndRefundValue(_owner);
+
+    // 2) Destroy token
+    emit Transfer(
```

Comment:
Hmm. I'm not seeing the value in leaving the owner holding an invalid key...unless this leads into a use-case I'm not seeing?
 I would be in favour of making the assumption here that someone cancelling their key doesn't want it anymore, so lets clean house a bit. If there's an exception to this, they can always buy another (but I do think that would be the exception).

Reply:
_If_ cancelling a key means"destroying" it and we transfer to `address(0)`, we should probably modify `ownerByTokenId` for the token as well...?

Reply:
Also, (and I may be missing something here) while this emits the `Transfer` event, it doesn't actually transfer the token. Our `transferFrom` function won't allow this because of `require(_recipient != address(0), "Can't Transfer to 0x Address");`
Should we consider adding a `burn` function to achieve this?

Reply:
The Transfer event to address(0) is the ERC721 standard way of destroying a Key.  I'm interpreting making a Key invalid to be the same as a destroy.  We could remove this and just lean on the cancel event.

Interesting point on cleanup.  If we do clear the data and ownerByTokeId, we should do the same for transferFrom.

Reply:
From ERC721:

```
    /// @dev This emits when ownership of any NFT changes by any mechanism.
    ///  This event emits when NFTs are created (`from` == 0) and destroyed
    ///  (`to` == 0). Exception: during contract creation, any number of NFTs
    ///  may be created and assigned without emitting Transfer. At the time of
    ///  any transfer, the approved address for that NFT (if any) is reset to none.
    event Transfer(address indexed _from, address indexed _to, uint256 indexed _tokenId);
```
"Also note that a NFTs MAY become invalid (be destroyed)."
"Creating of NFTs ("minting") and destruction NFTs ("burning") is not included in the specification. Your contract may implement these by other means. Please see the event documentation for your responsibilities when creating or destroying NFTs."

Reply:
So I am thinking we should have both a "cancel" (basically get a refund) and a "destroy" distinctly.

 I know I sound like a broken record but it will be very important down the line to know that someone previsouly owned a key (even though that this weak on the sybil attack side)

Reply:
I understand that the `transfer` event must emit "when ownership of any NFT changes by any mechanism". I was under the impression that we still had to set the new owner of the token to `address(0)` in the lock contract for this event to reflect the actual state? That is, I don't think that emitting the event is enough on its own. I will take another look, as you may have already done this and I missed it.
I'm in favour of cleaning up the lock contract where possible, but I understand that we don't want to remove any addresses from the `owners` array.

Reply:
Done. Removed emit Transfer

Reply:
Don't we still need the `Transfer` event to be emitted though? I was more looking at what happens after we emit `Transfer`...
In our contract, `ownerByTokenId` for this token would still return the old owner (but should actually now return `address(0)` right?)